### DISCUSSION TITLE
Netplay Accessibility Improvements

### DISCUSSION THREAD

**Discussion by assemblaj** [ikemen-engine/Ikemen-GO#1277] [Engine development]
Currently, IKEMEN requires VPNs or Streaming services such as Parsec (now defunct) in order to play matches against others over the internet.  My goal is to make IKEMEN's online more accessible, ideally such that users can connect to other players without having to use external software or leave the game itself. 

Currently I already have a Nakama server running.  In the past, there was no Go client libraries for Nakama (despite the server being written in the language), but within the past few months, [that's changed](https://github.com/ascii8/nakama-go). I've been experimenting with using Nakama to connect players without VPNs ([code found here](https://github.com/assemblaj/nakamatest)). I was able to successfully play from my laptop to my PC with no external software.  

How this works is that both players join the same Nakama room. A public STUN server is used to retrieve the peer's public IP, and each peer sends their public IP via Nakama.  That public IP is then passed to IKEMEN and then netplay continues as normal.  There's more complications to it than that (peers send messages to make users aware they're in the room, and also acknowledgements that they've received an IP), but the basic gist of it is that now that they have their opponents public IP, they can send packets directly to them. :-) 

How this could be used in the engine 
```go 
func (ni *NetInput) Accept() error {
	ip, port, err := nakamatest.GetPublicAddr("save2/onlineConfig.json", 7500)
	if err != nil {
		return err
	}
	sys.rollback.session.remoteIp = ip
	if ln, err := net.Listen("tcp", ":"+port); err != nil {
		return err
	} else {
		ni.ln = ln.(*net.TCPListener)
		ni.host = true
		ni.locIn, ni.remIn = ni.GetHostGuestRemap()
		go func() {
			ln := ni.ln
			defer ln.Close()     // Ensure the listener is closed when the function ends
			for ni.conn == nil { // Keep accepting connections until we find one from the right IP
				if conn, err := ln.AcceptTCP(); err == nil && conn.RemoteAddr().(*net.TCPAddr).IP.String() == ip {
					ni.conn = conn
					break // Exit the loop once we find the right connection
				}
			}
		}()
	}
	return nil
}

func (ni *NetInput) Connect() {
	ni.host = false
	ni.remIn, ni.locIn = ni.GetHostGuestRemap()
	go func() {
		ip, port, err := nakamatest.GetPublicAddr("save2/onlineConfig.json", 7500)
		sys.rollback.session.host = ip
		if err != nil {
			log.Fatal(err)
		}
		if conn, err := net.Dial("tcp", ip+":"+port); err == nil {
			ni.conn = conn.(*net.TCPConn)
		}
	}()
}
````

A big issue is adjusting the UI to support this feature, something that I cannot do. Right now I have a few ideas on how we can repurpose the current netplay UI for lobbies:

- On the host:
   - Selecting Host Game could lead you to a lobby name entry screen based on the current New Address screen
   - Then that could take you to the "waiting for connections from Player 2" screen 
- On the guest:
   - Selecting Join Game could dynamically populate a menu with lobbies (just like it currently loads hosts from config.json) 

Obviously other logic could be built on top of this (password protected lobbies, etc etc) but the biggest bottleneck to features like this is UI because I can't do that myself. 

Let me know if you have any ideas or comments.

**Comment by Lazin3ss** [comment]
I wonder if it's necessary to create another config file for netplay, as opposed to having everything inside config.json. Otherwise, great work, I'd like to help you design an UI for this feature in the future. I have experience with these kind of works in the engine (see: [battleplan.lua](https://github.com/Lazin3ss/OpenMK/blob/main/external/mods/battleplan.lua))
